File: bindings/ruby/lib/tasker_core/execution/command_client.rb
# frozen_string_literal: true

require 'logger'
require_relative '../types/execution_types'

module TaskerCore
  module Execution
    # Ruby wrapper for Rust-backed CommandClient
    #
    # This class provides a Ruby-friendly interface to the Rust-backed command client,
    # eliminating Ruby socket dependencies while maintaining API compatibility.
    #
    # @example Basic usage
    #   client = CommandClient.new(host: 'localhost', port: 8080)
    #   client.connect
    #
    #   response = client.register_worker(
    #     worker_id: 'ruby_worker_1',
    #     max_concurrent_steps: 10,
    #     supported_namespaces: ['orders', 'payments'],
    #     step_timeout_ms: 30000,
    #     supports_retries: true,
    #     language_runtime: 'ruby',
    #     version: RUBY_VERSION
    #   )
    #
    #   client.disconnect
    #
    class CommandClient
      # Default configuration
      DEFAULT_HOST = 'localhost'
      DEFAULT_PORT = 8080
      DEFAULT_TIMEOUT = 30
      DEFAULT_CONNECT_TIMEOUT = 5
      DEFAULT_NAMESPACE = 'default'

      attr_reader :host, :port, :timeout, :logger, :rust_client, :connected

      def initialize(host: nil, port: nil, timeout: nil)
        @config = TaskerCore::Config.instance.effective_config
        @host = host || @config.dig("command_backplane", "server", "host") || DEFAULT_HOST
        @port = port || @config.dig("command_backplane", "server", "port") || DEFAULT_PORT
        @timeout = timeout || @config.dig("command_backplane", "server", "timeout") || DEFAULT_TIMEOUT
        @logger = TaskerCore::Logging::Logger.instance
        @connected = false

        # Create Rust-backed command client
        client_config = {
          host: @host,
          port: @port,
          timeout_seconds: @timeout,
          connect_timeout_seconds: DEFAULT_CONNECT_TIMEOUT,
          namespace: DEFAULT_NAMESPACE
        }

        @rust_client = TaskerCore::CommandClient.new_with_config(client_config)
      end

      # Establish connection to Rust TCP executor
      #
      # @return [Boolean] true if connection successful
      # @raise [ConnectionError] if connection fails
      def connect
        return true if connected?

        begin
          success = @rust_client.connect
          if success
            @connected = true
            logger.info("Connected to Rust TCP executor at #{@host}:#{@port} via Rust client")
            true
          else
            raise ConnectionError, "Rust client connection failed"
          end
        rescue StandardError => e
          logger.error("Failed to connect to #{@host}:#{@port}: #{e.message}")
          @connected = false
          raise ConnectionError, "Failed to connect: #{e.message}"
        end
      end

      # Close connection to TCP executor
      def disconnect
        return unless @rust_client

        begin
          @rust_client.disconnect
          @connected = false
          logger.info("Disconnected from Rust TCP executor")
        rescue StandardError => e
          logger.warn("Error disconnecting Rust client: #{e.message}")
        end
      end

      # Check if client is connected
      #
      # @return [Boolean] true if connected
      def connected?
        @connected && @rust_client&.connected?
      end

      # Register a Ruby worker with the Rust orchestration system
      #
      # @param worker_id [String] Unique identifier for this worker
      # @param max_concurrent_steps [Integer] Maximum number of concurrent steps this worker can handle
      # @param supported_namespaces [Array<String>] List of namespaces this worker supports
      # @param step_timeout_ms [Integer] Timeout for individual steps in milliseconds
      # @param supports_retries [Boolean] Whether this worker supports step retries
      # @param language_runtime [String] Runtime identifier (usually 'ruby')
      # @param version [String] Runtime version
      # @param custom_capabilities [Hash] Additional worker capabilities
      # @param supported_tasks [Array<Hash>, nil] Task handler configurations with step templates
      # @return [TaskerCore::Types::ExecutionTypes::WorkerRegistrationResponse] Typed response from Rust orchestrator
      # @raise [NotConnectedError] if not connected
      # @raise [CommandError] if command fails
      def register_worker(worker_id:, max_concurrent_steps:, supported_namespaces: [DEFAULT_NAMESPACE],
                         step_timeout_ms:, supports_retries:, language_runtime:,
                         version:, custom_capabilities: {}, supported_tasks: nil)
        ensure_connected!

        options = {
          worker_id: worker_id,
          max_concurrent_steps: max_concurrent_steps,
          supported_namespaces: supported_namespaces,
          step_timeout_ms: step_timeout_ms,
          supports_retries: supports_retries,
          language_runtime: language_runtime,
          version: version,
          custom_capabilities: custom_capabilities
        }
        
        # Add supported_tasks if provided
        if supported_tasks && !supported_tasks.empty?
          options[:supported_tasks] = supported_tasks
          logger.debug("Including #{supported_tasks.size} supported tasks in worker registration")
        end

        begin
          raw_response = @rust_client.register_worker(options)
          logger.debug("Worker registration successful: #{worker_id}")
          
          # Convert raw response to typed WorkerRegistrationResponse
          TaskerCore::Types::ExecutionTypes::ResponseFactory.create_response(raw_response)
        rescue StandardError => e
          logger.error("Worker registration failed: #{e.message}")
          raise CommandError, "Worker registration failed: #{e.message}"
        end
      end

      # Send heartbeat to maintain worker connection
      #
      # @param worker_id [String] Worker identifier
      # @param current_load [Integer] Current number of steps being processed
      # @param system_stats [Hash, nil] Optional system statistics
      # @return [TaskerCore::Types::ExecutionTypes::HeartbeatResponse] Typed response from Rust orchestrator
      def send_heartbeat(worker_id:, current_load:, system_stats: nil)
        ensure_connected!

        options = {
          worker_id: worker_id,
          current_load: current_load,
          system_stats: system_stats
        }

        begin
          raw_response = @rust_client.send_heartbeat(options)
          logger.debug("Heartbeat successful for worker: #{worker_id}")
          
          # Convert raw response to typed HeartbeatResponse
          TaskerCore::Types::ExecutionTypes::ResponseFactory.create_response(raw_response)
        rescue StandardError => e
          logger.error("Heartbeat failed: #{e.message}")
          raise CommandError, "Heartbeat failed: #{e.message}"
        end
      end

      # Unregister worker from orchestration system
      #
      # @param worker_id [String] Worker identifier
      # @param reason [String] Reason for unregistration
      # @return [TaskerCore::Types::ExecutionTypes::WorkerUnregistrationResponse] Typed response from Rust orchestrator
      def unregister_worker(worker_id:, reason: 'Client shutdown')
        ensure_connected!

        options = {
          worker_id: worker_id,
          reason: reason
        }

        begin
          raw_response = @rust_client.unregister_worker(options)
          logger.debug("Worker unregistration successful: #{worker_id}")
          
          # Convert raw response to typed WorkerUnregistrationResponse
          TaskerCore::Types::ExecutionTypes::ResponseFactory.create_response(raw_response)
        rescue StandardError => e
          logger.error("Worker unregistration failed: #{e.message}")
          raise CommandError, "Worker unregistration failed: #{e.message}"
        end
      end

      # Send health check command
      #
      # @param diagnostic_level [String] Level of diagnostic information ('Basic', 'Detailed', 'Full')
      # @return [TaskerCore::Types::ExecutionTypes::HealthCheckResponse] Typed health check response
      def health_check(diagnostic_level: 'Basic')
        ensure_connected!

        begin
          raw_response = @rust_client.health_check(diagnostic_level)
          logger.debug("Health check successful")
          
          # Convert raw response to typed HealthCheckResponse
          TaskerCore::Types::ExecutionTypes::ResponseFactory.create_response(raw_response)
        rescue StandardError => e
          logger.error("Health check failed: #{e.message}")
          raise CommandError, "Health check failed: #{e.message}"
        end
      end

      # Send TryTaskIfReady command to check if task has ready steps and request batch creation
      #
      # @param task_id [Integer] Task ID to check for readiness
      # @return [TaskerCore::Types::ExecutionTypes::TaskReadinessResponse] Typed task readiness response
      def try_task_if_ready(task_id)
        ensure_connected!
        raise ArgumentError, "task_id must be an integer" unless task_id.is_a?(Integer)

        begin
          raw_response = @rust_client.try_task_if_ready(task_id)
          logger.debug("TryTaskIfReady command sent for task #{task_id}")
          
          # Convert raw response to typed TaskReadinessResponse
          TaskerCore::Types::ExecutionTypes::ResponseFactory.create_response(raw_response)
        rescue StandardError => e
          logger.error("TryTaskIfReady command failed for task #{task_id}: #{e.message}")
          raise CommandError, "TryTaskIfReady failed: #{e.message}"
        end
      end

      # Send InitializeTask command to create a new task
      #
      # @param task_request [Hash] Task request data
      # @return [TaskerCore::Types::ExecutionTypes::InitializeTaskResponse] Typed initialize task response
      def initialize_task(task_request)
        ensure_connected!
        raise ArgumentError, "task_request is required" unless task_request

        begin
          raw_response = @rust_client.initialize_task(task_request)
          logger.debug("InitializeTask command sent")
          
          # Convert raw response to typed response
          TaskerCore::Types::ExecutionTypes::ResponseFactory.create_response(raw_response)
        rescue StandardError => e
          logger.error("InitializeTask command failed: #{e.message}")
          raise CommandError, "InitializeTask failed: #{e.message}"
        end
      end

      # Get connection information
      #
      # @return [Hash] Connection information
      def connection_info
        return { connected: false } unless @rust_client

        begin
          info = @rust_client.connection_info
          # Return as-is since this is connection metadata, not a command response
          info.is_a?(Hash) ? info.transform_keys(&:to_sym) : info
        rescue StandardError => e
          logger.warn("Failed to get connection info: #{e.message}")
          { connected: false, error: e.message }
        end
      end

      private

      # Ensure client is connected
      #
      # @raise [NotConnectedError] if not connected
      def ensure_connected!
        return if connected?
        
        # Auto-reconnect for singleton pattern
        @logger.info "🔄 SINGLETON: Auto-reconnecting CommandClient for command execution"
        begin
          connect
          @logger.info "✅ SINGLETON: Auto-reconnection successful"
        rescue StandardError => e
          error_msg = "SINGLETON: Auto-reconnection failed: #{e.message}"
          @logger.error error_msg
          raise NotConnectedError, error_msg
        end
      end
    end

    # Custom error classes
    class CommandClientError < StandardError; end
    class ConnectionError < CommandClientError; end
    class NotConnectedError < CommandClientError; end
    class CommandError < CommandClientError; end
  end
end

File: bindings/ruby/ext/tasker_core/src/command_client.rs
//! # Ruby FFI Command Client Wrapper
//!
//! Magnus wrapper over SharedCommandClient from src/ffi/shared/command_client.rs
//! This provides Ruby FFI compatibility while delegating all operations to the shared
//! command client, eliminating Ruby socket dependencies and providing better integration
//! with our command pattern architecture.

use crate::context::{json_to_ruby_value, ruby_value_to_json};
use magnus::error::Result as MagnusResult;
use magnus::{function, method, Error, Module, Object, RModule, Value};
use std::sync::Arc;
use tasker_core::ffi::shared::command_client::{SharedCommandClient, CommandClientConfig, execute_async};
use tokio::sync::RwLock;
use tracing::{debug, error, info};

/// **RUBY FFI COMMAND CLIENT**: Magnus wrapper over SharedCommandClient
///
/// Provides Ruby FFI compatibility while delegating all operations to the shared
/// command client architecture, eliminating Ruby socket dependencies.
#[magnus::wrap(class = "TaskerCore::CommandClient")]
pub struct CommandClient {
    shared_client: Arc<RwLock<SharedCommandClient>>,
}

impl CommandClient {
    /// Create new Ruby command client wrapping shared client
    pub fn new() -> MagnusResult<Self> {
        info!("🔧 Ruby FFI: Creating CommandClient - delegating to shared client");

        let shared_client = SharedCommandClient::new_default();

        Ok(Self {
            shared_client: Arc::new(RwLock::new(shared_client)),
        })
    }

    /// Create new Ruby command client with custom configuration
    pub fn new_with_config(options: Value) -> MagnusResult<Self> {
        info!("🔧 Ruby FFI: Creating CommandClient with custom config");

        // Convert Ruby options to configuration
        let options_json = ruby_value_to_json(options).map_err(|e| {
            Error::new(
                magnus::exception::runtime_error(),
                format!("Failed to convert options: {e}"),
            )
        })?;

        let host = options_json
            .get("host")
            .and_then(|v| v.as_str())
            .unwrap_or("localhost")
            .to_string();
        let port = options_json
            .get("port")
            .and_then(|v| v.as_u64())
            .unwrap_or(8080) as u16;
        let timeout_seconds = options_json
            .get("timeout_seconds")
            .and_then(|v| v.as_u64())
            .unwrap_or(30) as u32;
        let connect_timeout_seconds = options_json
            .get("connect_timeout_seconds")
            .and_then(|v| v.as_u64())
            .unwrap_or(5) as u32;

        let config = CommandClientConfig {
            host,
            port,
            timeout_seconds,
            connect_timeout_seconds,
            read_timeout_seconds: 10,
            default_namespace: options_json
                .get("namespace")
                .and_then(|v| v.as_str())
                .unwrap_or("default")
                .to_string(),
        };

        let shared_client = SharedCommandClient::new(config);

        Ok(Self {
            shared_client: Arc::new(RwLock::new(shared_client)),
        })
    }

    /// Connect to the Rust TCP executor
    pub fn connect(&self) -> MagnusResult<bool> {
        debug!("🔧 Ruby FFI: connect() - delegating to shared client");

        let client = self.shared_client.clone();
        let result = execute_async(async move {
            let mut client_guard = client.write().await;
            client_guard.connect().await
        });

        match result {
            Ok(connected) => Ok(connected),
            Err(e) => Err(Error::new(
                magnus::exception::runtime_error(),
                format!("Connection failed: {e}"),
            )),
        }
    }

    /// Disconnect from the TCP executor
    pub fn disconnect(&self) -> MagnusResult<bool> {
        debug!("🔧 Ruby FFI: disconnect() - delegating to shared client");

        let client = self.shared_client.clone();
        let result = execute_async(async move {
            let mut client_guard = client.write().await;
            client_guard.disconnect().await
        });

        match result {
            Ok(_) => Ok(true),
            Err(e) => {
                error!("Disconnect failed: {}", e);
                Ok(false)
            }
        }
    }

    /// Check if client is connected
    pub fn connected(&self) -> MagnusResult<bool> {
        let client = self.shared_client.clone();
        let result = execute_async(async move {
            let client_guard = client.read().await;
            client_guard.is_connected().await
        });

        Ok(result)
    }

    /// Register a worker with the server
    pub fn register_worker(&self, options: Value) -> MagnusResult<Value> {
        debug!("🔧 Ruby FFI: register_worker() - delegating to shared client");

        // Convert Ruby options to parameters
        let options_json = ruby_value_to_json(options).map_err(|e| {
            Error::new(
                magnus::exception::runtime_error(),
                format!("Failed to convert worker options: {e}"),
            )
        })?;

        let worker_id = options_json
            .get("worker_id")
            .and_then(|v| v.as_str())
            .ok_or_else(|| {
                Error::new(
                    magnus::exception::runtime_error(),
                    "Worker options must include 'worker_id'",
                )
            })?
            .to_string();

        let max_concurrent_steps = options_json
            .get("max_concurrent_steps")
            .and_then(|v| v.as_u64())
            .unwrap_or(5) as u32;

        let supported_namespaces = options_json
            .get("supported_namespaces")
            .and_then(|v| v.as_array())
            .map(|arr| {
                arr.iter()
                    .filter_map(|v| v.as_str().map(|s| s.to_string()))
                    .collect()
            })
            .unwrap_or_else(|| vec!["default".to_string()]);

        let step_timeout_ms = options_json
            .get("step_timeout_ms")
            .and_then(|v| v.as_u64())
            .unwrap_or(30000) as u32;

        let supports_retries = options_json
            .get("supports_retries")
            .and_then(|v| v.as_bool())
            .unwrap_or(true);

        let language_runtime = options_json
            .get("language_runtime")
            .and_then(|v| v.as_str())
            .unwrap_or("ruby")
            .to_string();

        let version = options_json
            .get("version")
            .and_then(|v| v.as_str())
            .unwrap_or("3.0.0")
            .to_string();

        let custom_capabilities = options_json
            .get("custom_capabilities")
            .and_then(|v| v.as_object())
            .map(|obj| {
                obj.iter()
                    .map(|(k, v)| (k.clone(), v.clone()))
                    .collect()
            })
            .unwrap_or_else(|| std::collections::HashMap::new());

        let client = self.shared_client.clone();
        let result = execute_async(async move {
            let client_guard = client.read().await;
            client_guard
                .register_worker(
                    worker_id,
                    max_concurrent_steps,
                    supported_namespaces,
                    step_timeout_ms,
                    supports_retries,
                    language_runtime,
                    version,
                    custom_capabilities,
                )
                .await
        });

        match result {
            Ok(response) => {
                let response_json = serde_json::Value::Object(
                    response.into_iter().collect()
                );
                json_to_ruby_value(response_json).map_err(|e| {
                    Error::new(
                        magnus::exception::runtime_error(),
                        format!("Failed to convert response: {e}"),
                    )
                })
            }
            Err(e) => Err(Error::new(
                magnus::exception::runtime_error(),
                format!("Worker registration failed: {e}"),
            )),
        }
    }

    /// Send heartbeat to server
    pub fn send_heartbeat(&self, options: Value) -> MagnusResult<Value> {
        debug!("🔧 Ruby FFI: send_heartbeat() - delegating to shared client");

        // Convert Ruby options to parameters
        let options_json = ruby_value_to_json(options).map_err(|e| {
            Error::new(
                magnus::exception::runtime_error(),
                format!("Failed to convert heartbeat options: {e}"),
            )
        })?;

        let worker_id = options_json
            .get("worker_id")
            .and_then(|v| v.as_str())
            .ok_or_else(|| {
                Error::new(
                    magnus::exception::runtime_error(),
                    "Heartbeat options must include 'worker_id'",
                )
            })?
            .to_string();

        let current_load = options_json
            .get("current_load")
            .and_then(|v| v.as_u64())
            .unwrap_or(0) as u32;

        let system_stats = options_json
            .get("system_stats")
            .and_then(|v| v.as_object())
            .map(|obj| {
                obj.iter()
                    .map(|(k, v)| (k.clone(), v.clone()))
                    .collect()
            });

        let client = self.shared_client.clone();
        let result = execute_async(async move {
            let client_guard = client.read().await;
            client_guard
                .send_heartbeat(worker_id, current_load, system_stats)
                .await
        });

        match result {
            Ok(response) => {
                let response_json = serde_json::Value::Object(
                    response.into_iter().collect()
                );
                json_to_ruby_value(response_json).map_err(|e| {
                    Error::new(
                        magnus::exception::runtime_error(),
                        format!("Failed to convert response: {e}"),
                    )
                })
            }
            Err(e) => Err(Error::new(
                magnus::exception::runtime_error(),
                format!("Heartbeat failed: {e}"),
            )),
        }
    }

    /// Unregister worker from server
    pub fn unregister_worker(&self, options: Value) -> MagnusResult<Value> {
        debug!("🔧 Ruby FFI: unregister_worker() - delegating to shared client");

        // Convert Ruby options to parameters
        let options_json = ruby_value_to_json(options).map_err(|e| {
            Error::new(
                magnus::exception::runtime_error(),
                format!("Failed to convert unregister options: {e}"),
            )
        })?;

        let worker_id = options_json
            .get("worker_id")
            .and_then(|v| v.as_str())
            .ok_or_else(|| {
                Error::new(
                    magnus::exception::runtime_error(),
                    "Unregister options must include 'worker_id'",
                )
            })?
            .to_string();

        let reason = options_json
            .get("reason")
            .and_then(|v| v.as_str())
            .map(|s| s.to_string());

        let client = self.shared_client.clone();
        let result = execute_async(async move {
            let client_guard = client.read().await;
            client_guard.unregister_worker(worker_id, reason).await
        });

        match result {
            Ok(response) => {
                let response_json = serde_json::Value::Object(
                    response.into_iter().collect()
                );
                json_to_ruby_value(response_json).map_err(|e| {
                    Error::new(
                        magnus::exception::runtime_error(),
                        format!("Failed to convert response: {e}"),
                    )
                })
            }
            Err(e) => Err(Error::new(
                magnus::exception::runtime_error(),
                format!("Worker unregistration failed: {e}"),
            )),
        }
    }

    /// Send health check to server
    pub fn health_check(&self, diagnostic_level: Option<String>) -> MagnusResult<Value> {
        debug!("🔧 Ruby FFI: health_check() - delegating to shared client");

        let client = self.shared_client.clone();
        let result = execute_async(async move {
            let client_guard = client.read().await;
            client_guard.health_check(diagnostic_level).await
        });

        match result {
            Ok(response) => {
                let response_json = serde_json::Value::Object(
                    response.into_iter().collect()
                );
                json_to_ruby_value(response_json).map_err(|e| {
                    Error::new(
                        magnus::exception::runtime_error(),
                        format!("Failed to convert response: {e}"),
                    )
                })
            }
            Err(e) => Err(Error::new(
                magnus::exception::runtime_error(),
                format!("Health check failed: {e}"),
            )),
        }
    }

    /// Get connection information
    pub fn connection_info(&self) -> MagnusResult<Value> {
        let client = self.shared_client.clone();
        let result = execute_async(async move {
            let client_guard = client.read().await;
            client_guard.connection_info().await
        });

        let response_json = serde_json::Value::Object(
            result.into_iter().collect()
        );
        json_to_ruby_value(response_json).map_err(|e| {
            Error::new(
                magnus::exception::runtime_error(),
                format!("Failed to convert connection info: {e}"),
            )
        })
    }

    /// Send InitializeTask command to create a new task
    pub fn initialize_task(&self, task_request: Value) -> MagnusResult<Value> {
        debug!("🔧 Ruby FFI: initialize_task() - delegating to shared client");

        // Convert Ruby task_request to JSON
        let task_request_json = ruby_value_to_json(task_request).map_err(|e| {
            Error::new(
                magnus::exception::runtime_error(),
                format!("Failed to convert task_request: {e}"),
            )
        })?;

        let client = self.shared_client.clone();
        let result = execute_async(async move {
            let client_guard = client.read().await;
            client_guard.initialize_task(task_request_json).await
        });

        match result {
            Ok(response) => json_to_ruby_value(response).map_err(|e| {
                Error::new(
                    magnus::exception::runtime_error(),
                    format!("Failed to convert response: {e}"),
                )
            }),
            Err(e) => Err(Error::new(
                magnus::exception::runtime_error(),
                format!("InitializeTask failed: {e}"),
            )),
        }
    }

    /// Send TryTaskIfReady command to check if task has ready steps
    pub fn try_task_if_ready(&self, task_id: i64) -> MagnusResult<Value> {
        debug!("🔧 Ruby FFI: try_task_if_ready() - delegating to shared client");

        let client = self.shared_client.clone();
        let result = execute_async(async move {
            let client_guard = client.read().await;
            client_guard.try_task_if_ready(task_id).await
        });

        match result {
            Ok(response) => json_to_ruby_value(response).map_err(|e| {
                Error::new(
                    magnus::exception::runtime_error(),
                    format!("Failed to convert response: {e}"),
                )
            }),
            Err(e) => Err(Error::new(
                magnus::exception::runtime_error(),
                format!("TryTaskIfReady failed: {e}"),
            )),
        }
    }
}

/// Register the CommandClient class with Ruby
pub fn register_command_client(module: &RModule) -> MagnusResult<()> {
    let class = module.define_class("CommandClient", magnus::class::object())?;

    // Constructor methods
    class.define_singleton_method("new", function!(CommandClient::new, 0))?;
    class.define_singleton_method(
        "new_with_config",
        function!(CommandClient::new_with_config, 1),
    )?;

    // Connection methods
    class.define_method("connect", method!(CommandClient::connect, 0))?;
    class.define_method("disconnect", method!(CommandClient::disconnect, 0))?;
    class.define_method("connected?", method!(CommandClient::connected, 0))?;

    // Worker operations
    class.define_method(
        "register_worker",
        method!(CommandClient::register_worker, 1),
    )?;
    class.define_method(
        "send_heartbeat",
        method!(CommandClient::send_heartbeat, 1),
    )?;
    class.define_method(
        "unregister_worker",
        method!(CommandClient::unregister_worker, 1),
    )?;

    // Task operations
    class.define_method(
        "initialize_task",
        method!(CommandClient::initialize_task, 1),
    )?;
    class.define_method(
        "try_task_if_ready",
        method!(CommandClient::try_task_if_ready, 1),
    )?;

    // Health and info
    class.define_method("health_check", method!(CommandClient::health_check, 1))?;
    class.define_method(
        "connection_info",
        method!(CommandClient::connection_info, 0),
    )?;

    Ok(())
}